docs/turbo-native-modules.md

```diff
@@ -0,0 +1,255 @@
+
+Lets work through each of these steps by building an example Turbo Native Module.
+
+# Native Persistent Storage
+
+This guide will show you how to write an implementation of the [Web Storage API](https://html.spec.whatwg.org/multipage/webstorage.html#dom-localstorage-dev): `localStorage`. The API is relatable to a React developer who might be writing application code on your project.
+
+To make this work on mobile, we need to use 2 platform APIS:
+
+- iOS: [NSUserDefaults](https://developer.apple.com/documentation/foundation/nsuserdefaults), and
+- Android: [SharedPreferences](https://developer.android.com/reference/android/content/SharedPreferences).
+
+## 1. Declare Typed Specification
+
+React Native provides tooling called [Codegen](/the-new-architecture/what-is-codegen.md), which will take a specification written in TypeScript or Flow and generate platform specific code for iOS and Android. The specification declares the methods and data types that will pass back and forth between your native code and the React Native JavaScript runtime. We call your specification and the native code you write to work with the Codegen interfaces a Turbo Native Module.
+
+:::info
+You can see all of the types you can use in your specification and the native types that are generated in the [Appendix](/appendix.md) documentation.
+:::
+
+Here is an implementation of the `localStorage` specification:\*
+
+<Tabs groupId="language" queryString defaultValue={constants.defaultJavaScriptSpecLanguage} values={constants.javaScriptSpecLanguages}>
+<TabItem value="typescript">
+
+```typescript
+import {TurboModule, TurboModuleRegistry} from 'react-native';
+
+export interface Spec extends TurboModule {
+  setItem(value: string, key: string): void;
+  getItem(key: string): string | null;
+  removeItem(key: string): void;
+  clear(): void;
+}
+
+export default TurboModuleRegistry.getEnforcing<Spec>(
+  'NativeLocalStorage',
+) as Spec;
+```
+
+</TabItem>
+<TabItem value="flow">
+
+```flow
+import {TurboModule, TurboModuleRegistry} from 'react-native';
+
+export interface Spec extends TurboModule {
+  setItem(value: string, key: string): void;
+  getItem(key: string): ?string;
+  removeItem(key: string): void;
+  clear(): void;
+}
+```
+
+</TabItem>
+</Tabs>
+
+## 2. Write Application Code using the Turbo Native Module
+
+Next you’ll need to put your application developer hat on and write some JavaScript. Using `NativeLocalStorage`, here’s a modified `App.tsx` that includes some text we want persisted, an input field and some buttons to update this value. Since we’re application developers, lets added add dark mode support for extra polish:
+
+The `TurboModuleRegistry` supports 2 modes of retreiving a Turbo Native Module:
+
+- `get<T>(name: string): T | null` which will return `null` if the Turbo Native Module is unavailable. This is typically a concern for teams doing fancy things where they ship the JS and native code separately.
```

Comment:
I don't like that we don't provide some kind of explanation for the difference.  That's about as simple as I could make it without talking about shipping JS / Native separately / etc...

Reply:
but the "this is typically a concern for teams doing fancy things" is not a good explanation. Users can be upset as they might think that "So my team is not doing things fancy enough?".

Either we provide a technical explanation or we do not. Actually, the default use case is to use `get<T>`, as it is the safest approach (it does not crash the app and TS forces you to do the null-check).

`getEnforcing<T>` should be used only if you are super sure that the turbomodule will exists, otherwise it will crash the app. Consider that I believe that the code might return `null` also in case the initialization of the TM fails, for example if it has to use an external resource which is not available.